contracts/RootChain.sol

```diff
@@ -326,62 +327,88 @@ contract RootChain {
 
         // Make sure this exit is valid.
         require(output.amount > 0);
         require(exits[exitId].amount == 0);
 
+        InFlightExit storage inFlightExit = _getInFlightExit(_outputTx);
+        if (inFlightExit.exitStartTimestamp != 0) {
+            // Check if this output was piggybacked or exited in in-flight exit
+            require(!inFlightExit.exitMap.bitSet(oindex + 4) && !inFlightExit.exitMap.bitSet(oindex + 4 + 8));
+            // Prevent future piggybacks on this output
+            inFlightExit.exitMap = inFlightExit.exitMap.setBit(oindex + 4 + 8);
+        }
+
         // Make sure queue for this token exists.
         require(hasToken(output.token));
 
         // Determine the exit's priority.
         uint256 exitPriority = getStandardExitPriority(exitId, _outputId);
-        exitPriority = markStandard(exitPriority);
 
         // Insert the exit into the queue and update the exit mapping.
         PriorityQueue queue = PriorityQueue(exitsQueues[output.token]);
         queue.insert(exitPriority);
         exits[exitId] = Exit({
             owner: output.owner,
             token: output.token,
-            amount: output.amount
+            amount: output.amount,
+            position: _outputId
         });
 
         emit ExitStarted(output.owner, _outputId, output.amount, output.token);
     }
 
     /**
      * @dev Blocks a standard exit by showing the exiting output was spent.
-     * @param _outputId Identifier of the exiting output to challenge.
+     * @param _standardExitId Identifier of the standard exit to challenge.
      * @param _challengeTx RLP encoded transaction that spends the exiting output.
      * @param _inputIndex Which input to the challenging tx corresponds to the exiting output.
      * @param _challengeTxSig Signature from the exiting output owner over the spend.
      */
-    function challengeStandardExit(uint192 _outputId, bytes _challengeTx, uint256 _inputIndex, bytes _challengeTxSig)
+    function challengeStandardExit(uint192 _standardExitId, bytes _challengeTx, uint256 _inputIndex, bytes _challengeTxSig)
         public
     {
         // Check that the output is being used as an input to the challenging tx.
         uint256 inputId = _challengeTx.getInputId(_inputIndex);
-        require(inputId == _outputId);
-
-        uint192 exitId = getStandardExitId(_outputId);
+        require(inputId == exits[_standardExitId].position);
 
+        // Check if exit exists.
+        address owner = exits[_standardExitId].owner;
         // Check that the challenging tx is signed by the output's owner.
-        address owner = exits[exitId].owner;
         bytes32 txHash = keccak256(_challengeTx);
         require(owner == ECRecovery.recover(txHash, _challengeTxSig));
 
+        processChallengeStandardExit(inputId, _standardExitId);
+    }
+
+    function maybeChallengeStandardExitOnInput(uint256 _outputId, bytes _txbytes)
```

Comment:
despite internal, some brief comment explaining this function is required, hard to figure out based on name solely. Or maybe there's a way to call it more self-explanatorily?

Reply:
done